rcl/src/rcl/rmw_implementation_identifier_check.c

```diff
@@ -146,12 +156,22 @@ rcl_ret_t rcl_rmw_implementation_identifier_check(void)
         actual_rmw_impl_id,
         RCL_RET_MISMATCHED_RMW_ID
       );
-      return RCL_RET_MISMATCHED_RMW_ID;
+      fail_state = RCL_RET_MISMATCHED_RMW_ID;
+      goto fail_cleanup;
     }
     // Free the memory now that all checking has passed.
     allocator.deallocate((char *)expected_rmw_impl, allocator.state);
   }
   return RCL_RET_OK;
+
+fail_cleanup:
+  if (expected_rmw_requires_free) {
+    allocator.deallocate((char *)expected_rmw_impl, allocator.state);
+  }
+  if (asserted_rmw_requires_free) {
+    allocator.deallocate((char *)asserted_rmw_impl, allocator.state);
+  }
+  return fail_state;
```

Comment:
You can share the normal cleanup code with the error cleanup code to make things easier.
```suggestion
  }
  fail_state = RCL_RET_OK;  // fail_state should be renamed as ret for clarity
// fallthrough
cleanup:  // fail_cleanup is now cleanup
  if (expected_rmw_impl) {
    allocator.deallocate((char *)expected_rmw_impl, allocator.state);
  }
  if (asserted_rmw_impl) {
    allocator.deallocate((char *)asserted_rmw_impl, allocator.state);
  }
  return fail_state;
```

Reply:
I don't understand why github renders the suggestion so badly :joy:

Reply:
Thanks for the suggestion! I added those with 9e0bed9